File: goldencheck/profilers/nullability.py
"""Nullability profiler — detects required vs. optional columns."""
from __future__ import annotations
import polars as pl
from goldencheck.models.finding import Finding, Severity
from goldencheck.profilers.base import BaseProfiler

class NullabilityProfiler(BaseProfiler):
    def profile(self, df: pl.DataFrame, column: str, *, context: dict | None = None) -> list[Finding]:
        findings: list[Finding] = []
        col = df[column]
        total = len(col)
        null_count = col.null_count()
        null_pct = null_count / total if total > 0 else 0
        if null_count == total:
            findings.append(Finding(severity=Severity.ERROR, column=column, check="nullability",
                message=f"Column is entirely null ({total} rows)", affected_rows=total,
                confidence=0.99))
        elif null_count == 0 and total >= 10:
            # 0 nulls in 1000+ rows → 0.95; 0 nulls <50 → 0.5; else 0.7
            if total >= 1000:
                confidence = 0.95
            elif total < 50:
                confidence = 0.5
            else:
                confidence = 0.7
            findings.append(Finding(severity=Severity.INFO, column=column, check="nullability",
                message=f"0 nulls across {total} rows — likely required",
                confidence=confidence))
        elif null_pct > 0 and null_pct < 1:
            findings.append(Finding(severity=Severity.INFO, column=column, check="nullability",
                message=f"{null_count} nulls ({null_pct:.1%}) — column is optional", affected_rows=null_count,
                confidence=0.7))
        return findings

File: packages/goldencheck-js/src/core/profilers/nullability.ts
/**
 * Nullability profiler — detects required vs. optional columns.
 * Port of goldencheck/profilers/nullability.py.
 */

import type { TabularData } from "../data.js";
import { type Finding, Severity, makeFinding } from "../types.js";
import type { Profiler } from "./base.js";

export class NullabilityProfiler implements Profiler {
  profile(data: TabularData, column: string): Finding[] {
    const findings: Finding[] = [];
    const total = data.rowCount;
    const nullCount = data.nullCount(column);
    const nullPct = total > 0 ? nullCount / total : 0;

    if (nullCount === total) {
      findings.push(
        makeFinding({
          severity: Severity.ERROR,
          column,
          check: "nullability",
          message: `Column is entirely null (${total} rows)`,
          affectedRows: total,
          confidence: 0.99,
        }),
      );
    } else if (nullCount === 0 && total >= 10) {
      let confidence: number;
      if (total >= 1000) confidence = 0.95;
      else if (total < 50) confidence = 0.5;
      else confidence = 0.7;

      findings.push(
        makeFinding({
          severity: Severity.INFO,
          column,
          check: "nullability",
          message: `0 nulls across ${total} rows — likely required`,
          confidence,
        }),
      );
    } else if (nullPct > 0 && nullPct < 1) {
      const nonNullPct = 1.0 - nullPct;

      if (nonNullPct > 0.95 && total >= 100) {
        findings.push(
          makeFinding({
            severity: Severity.WARNING,
            column,
            check: "nullability",
            message: `${nullCount} nulls (${(nullPct * 100).toFixed(1)}%) in a ${(nonNullPct * 100).toFixed(1)}% non-null column — possible data quality issue`,
            affectedRows: nullCount,
            suggestion: "Verify whether these nulls are expected or indicate missing data",
            confidence: 0.75,
          }),
        );
      } else {
        const notable =
          nullPct > 0.8 || (total >= 100 && nullPct > 0.05);

        if (notable) {
          findings.push(
            makeFinding({
              severity: Severity.INFO,
              column,
              check: "nullability",
              message: `${nullCount} nulls (${(nullPct * 100).toFixed(1)}%) — column is optional`,
              affectedRows: nullCount,
              confidence: 0.7,
            }),
          );
        }
      }
    }

    return findings;
  }
}
